src/RA_StringUtils.h

```
@@ -25,65 +25,486 @@ _NODISCARD std::string Narrow(_In_ const std::string& wstr);
-        nNeeded = std::vswprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
+        Append(arg);
+        Append(std::forward<Ts>(args)...);
     }
-    va_end(pArgs);
-    assert(nNeeded >= 0);
-    assert(nNeeded < RSIZE_MAX/sizeof(CharT));
-    if (nNeeded >= 0)
+
+    template<typename T>
+    void AppendFormat(_UNUSED T arg, _UNUSED const std::string& sFormat)
     {
-        if (ra::to_unsigned(nNeeded) >= sFormatted.size())
+        // cannot use static_assert here because the code will get generated regardless of if its ever used
+        assert(!"Unsupported formatted type");
+    }
+
+    template<> 
+    void AppendFormat(float arg, const std::string& sFormat)
+    {
+        std::ostringstream oss;
+
+        if (sFormat.back() == 'f' || sFormat.back() == 'F')
         {
-            sFormatted.resize(ra::to_unsigned(nNeeded + 1));
-            int nCheck = 0;
-            va_start(pArgs, sFormat);
-            if constexpr (std::is_same_v<CharT, char>)
-                nCheck = std::vsnprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
-            else if constexpr (std::is_same_v<CharT, wchar_t>)
-                nCheck = std::vswprintf(sFormatted.data(), sFormatted.capacity(), sFormat, pArgs);
-            va_end(pArgs);
-            assert(nCheck == nNeeded);
+            int nIndex = sFormat.find('.');
+            if (nIndex != std::string::npos)
+            {
+                int nPrecision = atoi(sFormat.c_str() + nIndex + 1);
+                oss.precision(nPrecision);
+                oss << std::fixed;
+            }
+        }
+
+        oss << arg;
+        Append(oss.str());
+    }
+
+    template<>
+    void AppendFormat(int arg, const std::string& sFormat)
+    {
+        std::ostringstream oss;
+        if (sFormat.front() == '0')
+            oss << std::setfill('0');
+        int nDigits = atoi(sFormat.c_str());
```

I'm going to suggest using `std::stoi`, mainly because `atoi` fails silently. They basically do the same thing, but stoi is more secure.

---

done